Improve readmes (#1365)
balancer-labs/balancer-v2-monorepo: pkg/deployments/DEPLOYING.md

# Deploying Balancer Contracts

> The deployments package aims to provide a robust mechanism to run and verify deployments on multiple networks. This document describes rationale, what guarantees it provides, and how to create and run deployments.

Reviewer: ```suggestion
The deployments package aims to provide a robust mechanism to run and verify deployments on multiple networks. This document describes the rationale, what guarantees it provides, and how to create and run deployments.
```

## Overview

> Each deployment unit is called a 'task'. Tasks have a unique ID, which is often prefixed with the date in which they were first executed (to e.g. tell apart the deployment of Weighted Pools from April 2021, and the one in July 2022, which might feature widly different behavior).

Reviewer: ```suggestion
Each deployment unit is called a 'task'. Tasks have a unique ID, which is often prefixed with the date on which they were first executed (to e.g. differentiate the April 2021 deployment of Weighted Pools from the one deployed in July 2022, which might have wildly different behavior).
```

A task is made up of multiple components:

> - one or multiplle build information JSON files, located in the `build-info` directory.

Reviewer: ```suggestion
- one or multiple build information JSON files, located in the `build-info` directory.
```

> - an `input.ts` file, containing information on dependencies and deployment parameters (which might differ accross networks).

Reviewer: ```suggestion
- an `input.ts` file, containing information on dependencies and deployment parameters (which might differ across networks).
```

- the `output`, `abi` and `bytecode` directories, with JSON files containing the deployment addresses of the tasks' contracts in each network as well as additional information.
- optional fork tests in the `test` directory.

### Why `build-info`

When Hardhat compiles a Solidity project, in addition to the traditional contract artifacts it also produces a special file in the `artifacts/build-info` directory, which includes all of the compilation inputs (full source code, compiler version and settings, etc.) and all of the compilation outputs (creation bytecode, ABI, AST, etc.). By storing this file in the repository and using it as the source of truth for deployments, we fully break the dependency on the source code and are able to modify it and update it after a deployment is created, while guaranteeing that all deployments of a given task will be identical (since they share the same build information).

As part of our development process, we frequently alter the source code of contracts that have already been deployed, and keep no historical versions: the deployment data is fully stored in the build information file.

## Creating Tasks

> A new task is created by placing a new directory named after the task ID in the `tasks`, and populating the `build-info` directory and writing appropiate `input.ts` and `index.ts` files. The `output`, `abi` and `bytecode` directories are automatically generated and populated.

Reviewer: ```suggestion
A new task is created by placing a new directory named after the task ID in the `tasks` directory, and populating the `build-info` directory and writing appropriate `input.ts` and `index.ts` files. The `output`, `abi` and `bytecode` directories are automatically generated and populated by running the extract-artifacts script after the build-info is present.
```

> The build information should be generated on a clean commit, so that it can be verified by other parties. On the project directory, delete the `artifacts` directory and run `yarn hardhat compile` for a clean compilation. The file will be then be located at `artifacts/build-info`.

Reviewer: Do we need to explain about editing package.json to remove the part that normally deletes build-info on compilation? Or make a compile-with-buildinfo?

Author: `yarn hardhat compile` will do the right thing. I don't want to go into too much detail tbh - this is not necessarily a step-by-step guide. And the reason why we delete the buldinfos is entirely accidental.

Inputs and task instructions are plain TypeScript files that call appropriate functions - there's no DSL. The recommended way to write them is to copy the structure from the `input.ts` and `index.ts` files from a similar task, and then edit those as needed.

### Vyper Compilation

Hardhat does not generate build information for Vyper contracts - those must be created manually.

## Verifying Tasks

> There's three things to verify from a task:

Reviewer: ```suggestion
There are three things to verify from a task:
```

- the build information
- the inputs
- the deployment script

> The last two can be performed with simple manual inspection, that is, reviewing `input.ts` and `index.ts` and checking that inputs are sensible and constructor arguments provided correctly.

Reviewer: ```suggestion
The last two can be performed with simple manual inspection, that is, reviewing `input.ts` and `index.ts` and checking that they have sensible inputs and correct constructor arguments.
```

    "gasPrice": "auto",
    "gasMultiplier": 1,
    "accounts": ["0x-eoa-private-key"]
  }
}
```

### `deploy`

Once all prerequisites are met and the task is ready, its contracts can be deployed via the custom `deploy` Hardhat task, which can be run via:

```bash
$ yarn hardhat deploy --id <task-id> --network <network>
```

Deployment addresses will be automatically saved to the appropriate file in the tasks's `output` directory, and source code submittted to Etherscan for verification[^1].

Further runs of the task will not attempt to redeploy contracts for which an output already exists: use the `--force` flag to do a redeployment.

## Checking Deployments

The deployments package is able to verify that the addresses found in the `output` directories correspond exactly to a run of the task as described by `index.ts`, using the inputs at `input.ts` and the data from the `build-info` directory.

This can be done manually by running the `check-deployments` Hardhat task (`yarn hardhat check-deployments --network <network>`), but it is also done automatically on CI. Like deployment, [this also requires access to an RPC endpoint](#prerequisites).

## Running Tests

> Some tasks define fork tests, which consist of creating a local form from real network and some block, running the task on said fork, and then performing tests on top.

Reviewer: ```suggestion
Some tasks define fork tests, which consist of creating a local fork from a real network at a given block, running the task on said fork, and then performing tests.
```

Unlike regular deployments, this requires access to an _archive node_ instead of a regular RPC endpoint. [Alchemy](https://www.alchemy.com/) provides access to these on their free tier. They are configured the same way as [regular deployment endpoints](#prerequisites) (it is possible to simply use the archive endpoint for all purposes, including contract deployment).

> Once that is setup, fork tests can be run:

Reviewer: ```suggestion
Once that is set up, fork tests can be run:
```
